### PR TITLE
Sync Button

### LINKED ISSUE #33: [Syncing] Button
## Points to note
- Component in action - `Button.js`
- Github Urls to this component:
  * [neeto-ui/Button.js](https://github.com/bigbinary/neeto-ui/tree/67ee4a5188c5feccbeca2767107651fb24932764/lib/components/Button.js)
  * [NitroUI/Button.js](https://github.com/bigbinary/neeto-help-web/tree/0b11a7b0f98ac7c533df1e9ec37cf900b4cbc901/app/javascript/Admin/Components/NitroUI/Button.js)
- :green_circle: Green color signifies how that particular line is written in `neeto-ui` package
- :red_circle: Red color signifies how that particular line is written in `NitroUI` internals

## Sync Required or Not
- Are there significant changes for `Button.js` that needs to be synced to `neeto-ui` package from `NitroUI`? - **No**
- The `useRef` doesn't add any value from my understanding

## Diff
```diff
--- NitroUI/Button.js	2020-11-28 16:54:23.826521600 +0530
+++ neetoui-official/Button.js	2020-11-28 16:49:18.735162200 +0530
@@ -1,4 +1,4 @@
-import React, { useEffect, useRef } from "react";
+import React from "react";
 import classnames from "classnames";
 import { BeatLoader } from "react-spinners";
 import { Link } from "react-router-dom";
@@ -22,17 +22,10 @@
     dataTestId,
     size,
     href,
-    focus,
     ...otherProps
   } = props;
 
   let innerContent = label;
-  const buttonRef = useRef(null);
-  useEffect(() => {
-    if (focus) {
-      buttonRef.current && buttonRef.current.focus();
-    }
-  }, []);
 
   if (icon) {
     if (iconPosition === "right") {
@@ -66,13 +59,12 @@
 
   return (
     <Parent
-      ref={buttonRef}
       to={to}
       href={href}
       onClick={handleClick}
-      type={Parent === "button" ? type : null}
+      type={type}
       className={classnames(
-        [`nh-btn nh-btn--${style || "primary"}`, className],
+        [`nui-btn nui-btn--${style || "primary"}`, className],
         {
           "px-4 py-2": size === "large",
           "w-full justify-center py-2.5": fullWidth,
@@ -91,7 +83,7 @@
 
 const Loader = () => {
   return (
-    <div className="nh-btn--loader">
+    <div className="nui-btn--loader">
       <BeatLoader size={8} />
     </div>
   );
```

### PR DESCRIPTION
**Pull request by yedhink**
Fixes #33 

Merge this only after merging #63

@edwinbbu  _a
Would you be able to review and merge this PR? Thanks

cc: @goutham-subramanyam

### CODE REVIEW

File: lib/components/Button.js
Comment (goutham-subramanyam) [inline lib/components/Button.js:25]: We don't need the focus prop

**Review by goutham-subramanyam — CHANGES_REQUESTED** [review]
@yedhink _a

**Comment by edwinbbu** [comment]
@yedhink Lets not wait for #63 to get merged. #63 is blocked by another PR. Hence please make the necessary changes.

File: lib/components/Button.js
Comment (yedhink) [reply lib/components/Button.js:25]: @goutham-subramanyam 

If i'm understanding correctly, then we can remove this section too if we remove `focus` attr right:
https://github.com/bigbinary/neeto-help-web/blob/e962f3766ccb579b2f731035067ce57c819510c2/app/javascript/Admin/Components/NitroUI/Button.js#L32-L34

**Comment by yedhink** [comment]
@edwinbbu The only change to merge this before #63 would be to swap the package versions. I will make that change.

File: lib/components/Button.js
Comment (goutham-subramanyam) [reply lib/components/Button.js:25]: Yes

**Comment by yedhink** [comment]
@vinay0x _a

I need all of your opinions on this one. I have removed the `noop` function and made `onClick` prop `required`. I mean if we are creating a button, then it should have some click functionality associated with it, right?

In cases of using as submit button in forms and all, I think the developer can set the onClick to null explicitly where it's invoked at. This makes the code more clear rather than the logic being abstracted away for this particular attribute.

If everything looks good, can you also merge this PR? Thanks

cc: @edwinbbu @goutham-subramanyam

**Comment by edwinbbu** [comment]
> @vinay0x _a
> 
> I need all of your opinions on this one. I have removed the `noop` function and made `onClick` prop `required`. I mean if we are creating a button, then it should have some click functionality associated with it, right?
> 
> In cases of using as submit button in forms and all, I think the developer can set the onClick to null explicitly where it's invoked at. This makes the code more clear rather than the logic being abstracted away for this particular attribute.
> 
> If everything looks good, can you also merge this PR? Thanks
> 
> cc: @edwinbbu @goutham-subramanyam

@yedhink Button component doesn't always have `onClick`. Button component can act like a link also. In that case we provide `to` props.

**Comment by yedhink** [comment]
Ok. If setting `onClick` default to `() => {}` is advantageous or helps with writing less code, then let's use the `noop` function as default value. I will revert.

**Comment by yedhink** [comment]
@goutham-subramanyam _a

Would you be able to review and merge if everything looks good? Thanks

File: lib/components/Button.js
Comment (goutham-subramanyam) [inline lib/components/Button.js:29]: We don't need to add ref here. We just need to forward the ref like in Input.js

File: src/index.js
Comment (goutham-subramanyam) [inline src/index.js:30]: Why h-10 class ?

**Review by goutham-subramanyam — CHANGES_REQUESTED** [review]
@Swastikyadav _a

**Comment by edwinbbu** [comment]
@yedhink _a

**Comment by yedhink** [comment]
@goutham-subramanyam 
I'm confused. [Were these](https://github.com/bigbinary/neeto-ui/pull/66#pullrequestreview-544369872) questions that you wanted to ask Swastik? Or did you mean to ask me?

If it was meant for me:
- Sure I will update the ref to be like in `Input.js`
- The `h-10` class doesn't affect the component in any way. It's just to show an example of the Button component in action. Also the code snippet is directly taken from neeto-help-web.

**Comment by yedhink** [comment]
@goutham-subramanyam _a

I have made the review changes.

In my opinion, the previous usage without having `forwardRef` was fine looking and the code looked much more neater. But, i'm guessing there might be advantages for using `React.forwardRef` over simply passing the created ref as one among the props.

The question I have is, couldn't we just pass the created ref as say attribute `ref` to the `Button` component props? Wouldn't that allow passing multiple refs? Also since many of our components are class based, can we use `forwardRef` there? What are the advantages of using `forwardRef` over that strategy?

Also if you are wondering about the strategy that I had mentioned above, then [please do refer this](https://gist.github.com/gaearon/1a018a023347fe1c2476073330cc5509).

Also I [read this para from react official docs](https://reactjs.org/docs/forwarding-refs.html#note-for-component-library-maintainers). Please do check it out and review this PR and merge if everything looks good.

cc: @edwinbbu

**Review by goutham-subramanyam — APPROVED** [review]
Looks good. Merging.

**Pull request merged**
